apps/dashboard/src/components/primitives/field-editor/variable-plugin/utils.ts

```diff
@@ -0,0 +1,109 @@
+import { EditorView } from '@uiw/react-codemirror';
+import type { VariableMatch } from './types';
+
+/**
+ * Parses a variable match from the editor's content into structured data.
+ * This function is crucial for the variable pill system as it:
+ * 1. Extracts the position and content of variables like {{ subscriber.name | uppercase }}
+ * 2. Separates the base variable name from its modifiers (filters after |)
+ * 3. Provides the necessary information for rendering variable pills in the editor
+ *
+ * @example
+ * Input match for "{{ subscriber.name | uppercase }}"
+ * Returns:
+ * {
+ *   fullVariableName: "subscriber.name | uppercase",
+ *   variableName: "subscriber.name",
+ *   start: [match start index],
+ *   end: [match end index],
+ *   hasModifiers: true
+ * }
+ */
+export function parseVariable(match: RegExpExecArray): VariableMatch {
+  const start = match.index;
+  const end = start + match[0].length;
+  const fullVariableName = match[1].trim();
+  const parts = fullVariableName.split('|').map((part) => part.trim());
+  const variableName = parts[0];
+  const hasModifiers = parts.length > 1;
+
+  return {
+    fullVariableName,
+    variableName,
+    start,
+    end,
+    hasModifiers,
+  };
+}
+
+/**
+ * Handles backspace behavior for variables in the editor.
+ * When backspace is pressed near or within a variable, it removes the entire variable including its brackets.
+ */
+export function handleVariableBackspace(view: EditorView, pos: number, content: string): boolean {
```

Comment:
:bulp If I get it right, this ensures the pill is treated as a single character during keyboard interactions. Code mirror uses content-editable divs. So, to create pills that behave as a single character inside a content-editable div, there is a technique to nest non-content-editable spans.

See [here](https://jsbin.com/mohuyokena/edit?html,output). Since we fully control the rendering process with the view plugin, maybe it provides a non-JS solution to the problem.